ssdb-configuration-options.adoc

```diff
@@ -0,0 +1,560 @@
+are applicable with a `Redis` backend.
+
+== Overview
+
+=== High-availability with Master-Slave replication
+
+==== Normal Operation
+
+The method described in this document to implement master-slave
+replication makes use of http://www.keepalived.org/[Keepalived] to
+maintain a configured virtual IP address that is bound to the master
+`SSDB` instance under normal conditions.
+
+
+image:images/ssdb-master-slave-normal.png[]
+
+==== Fail-over
+
+If `Keepalived` detects a failure of the master instance, then the
+backup instance is promoted to master by re-binding the virtual IP
+address to the backup.
+
+
+image:images/ssdb-master-slave-failover.png[]
+
+
+=== High-availability with Master-Master replication
+
+This differs from master-slave replication in that both `SSDB`
+instances are online and accessible.  Each replicates changes from the
+other.  In the example set-up described later, we use
+http://www.haproxy.org/[HAProxy] as a front-end.  Keep in mind that,
+for production, you must use a proxy service that is, itself, highly-available.
+
+
+image:images/ssdb-master-master.png[]
+
+=== Horizontal Scaling via Multi-Master Configuration
+
+Normally, both `SSDB` and `Redis` contain the entire key-space in a
+single instance.  It is possible to front-end multiple instances using
+a service such as https://github.com/twitter/twemproxy[twemproxy].  It
+supports various hashing modes such that the data associated with a
+particular key is always stored on the same server.  This allows for
+horizontal scaling with very large installations.
+
+By configuring each `SSDB` instance in master-slave configuration, you
+get both horizontal scaling with high-availability.
+
+
+image:images/ssdb-multi-master.png[]
+
+
+== Master-Slave Replication
+
+
+One way to ensure that `SSDB` remains highly-available is to set-up
+master-slave replication and configure a system that will allow the
+backup `SSDB` instance to automatically take-over in the event that
+the primary instance goes down.  This document will describe one
+technique for accomplishing that goal.
+
+
+=== Overview
+
+`SSDB` will be installed on two servers.  It will be configured such
+that one server is the designated master and the other server is the
+designated slave, or backup, that will constantly replicate changes
+from the master server.
+
+`Keepalived` will also be installed on these two servers.  Each
+`Keepalived` instance will monitor the other.  In the event that the
+master server goes down, `Keepalived` will detect that failure and
+promote the slave, or backup, server to master.  `Keepalived` will
+maintain a https://en.wikipedia.org/wiki/Virtual_IP_address[Virtual
+IP address] bound to whichever server is the current master.
+
+The {product-name} will be configured such that the
+`zimbraEphemeralBackendURL` will bind to the _Virtual IP address_ that is
+being maintained by `Keepalived`.  Once the installation and
+configuration has been completed, follow the instructions in
```

Comment:
I would say "Once the installation and configuration of both the SSDB Master-Slave setup and {product-name} has been completed"